### ISSUE TITLE
When internet is down, library will eventually fail to connect without notifying client

### ISSUE THREAD

**Issue by adhawkins** [Teekeks/pyTwitchAPI#318]
I have a chat bot based on pytwitchapi. My internet was down for a few hours recently, and it was only when I happened to notice the bot wasn't working when I realised that it had failed to connect to the Twitch API (understandably) and gave up attempting to reconnect:

```
Jul 17 01:55:13 xcp-dev python[808]: Reached timeout for websocket receive, will attempt a reconnect
Jul 17 01:55:16 xcp-dev python[808]: connection attempt failed, retry in 0s...
Jul 17 01:55:20 xcp-dev python[808]: connection attempt failed, retry in 1s...
Jul 17 01:55:21 xcp-dev python[808]: connection attempt failed, retry in 2s...
Jul 17 01:55:23 xcp-dev python[808]: connection attempt failed, retry in 4s...
Jul 17 01:55:27 xcp-dev python[808]: connection attempt failed, retry in 8s...
Jul 17 01:55:39 xcp-dev python[808]: connection attempt failed, retry in 16s...
Jul 17 01:55:58 xcp-dev python[808]: connection attempt failed, retry in 32s...
Jul 17 01:56:33 xcp-dev python[808]: connection attempt failed, retry in 64s...
Jul 17 01:56:50 xcp-dev python[808]: Reached timeout for websocket receive, will attempt a reconnect
Jul 17 01:56:53 xcp-dev python[808]: connection attempt failed, retry in 0s...
Jul 17 01:56:57 xcp-dev python[808]: connection attempt failed, retry in 1s...
Jul 17 01:56:58 xcp-dev python[808]: connection attempt failed, retry in 2s...
Jul 17 01:57:00 xcp-dev python[808]: connection attempt failed, retry in 4s...
Jul 17 01:57:04 xcp-dev python[808]: connection attempt failed, retry in 8s...
Jul 17 01:57:15 xcp-dev python[808]: connection attempt failed, retry in 16s...
Jul 17 01:57:35 xcp-dev python[808]: connection attempt failed, retry in 32s...
Jul 17 01:57:40 xcp-dev python[808]: connection attempt failed, retry in 128s...
Jul 17 01:58:10 xcp-dev python[808]: connection attempt failed, retry in 64s...
Jul 17 01:59:17 xcp-dev python[808]: connection attempt failed, retry in 128s...
Jul 17 01:59:48 xcp-dev python[808]: Connection to chat websocket lost and unable to reestablish connection!
Jul 17 01:59:48 xcp-dev python[808]: Traceback (most recent call last):
Jul 17 01:59:48 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/aiohttp/client_ws.py", line 244, in receive
Jul 17 01:59:48 xcp-dev python[808]:     msg = await self._reader.read()
Jul 17 01:59:48 xcp-dev python[808]:           ^^^^^^^^^^^^^^^^^^^^^^^^^
Jul 17 01:59:48 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/aiohttp/streams.py", line 681, in read
Jul 17 01:59:48 xcp-dev python[808]:     return await super().read()
Jul 17 01:59:48 xcp-dev python[808]:            ^^^^^^^^^^^^^^^^^^^^
Jul 17 01:59:48 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/aiohttp/streams.py", line 640, in read
Jul 17 01:59:48 xcp-dev python[808]:     await self._waiter
Jul 17 01:59:48 xcp-dev python[808]: asyncio.exceptions.CancelledError
Jul 17 01:59:48 xcp-dev python[808]: During handling of the above exception, another exception occurred:
Jul 17 01:59:48 xcp-dev python[808]: Traceback (most recent call last):
Jul 17 01:59:48 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/twitchAPI/chat/__init__.py", line 913, in __task_receive
Jul 17 01:59:48 xcp-dev python[808]:     message = await self.__connection.receive(timeout=receive_timeout)
Jul 17 01:59:48 xcp-dev python[808]:               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jul 17 01:59:48 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/aiohttp/client_ws.py", line 243, in receive
Jul 17 01:59:48 xcp-dev python[808]:     async with async_timeout.timeout(timeout or self._receive_timeout):
Jul 17 01:59:48 xcp-dev python[808]:   File "/usr/lib/python3.11/asyncio/timeouts.py", line 98, in __aexit__
Jul 17 01:59:48 xcp-dev python[808]:     raise TimeoutError
Jul 17 01:59:48 xcp-dev python[808]: TimeoutError
Jul 17 01:59:48 xcp-dev python[808]: During handling of the above exception, another exception occurred:
Jul 17 01:59:48 xcp-dev python[808]: Traceback (most recent call last):
Jul 17 01:59:48 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/twitchAPI/chat/__init__.py", line 918, in __task_receive
Jul 17 01:59:48 xcp-dev python[808]:     await self._handle_base_reconnect()
Jul 17 01:59:48 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/twitchAPI/chat/__init__.py", line 956, in _handle_base_reconnect
Jul 17 01:59:48 xcp-dev python[808]:     await self.__connect(is_startup=False)
Jul 17 01:59:48 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/twitchAPI/chat/__init__.py", line 866, in __connect
Jul 17 01:59:48 xcp-dev python[808]:     raise TwitchBackendException('can\'t connect')
Jul 17 01:59:48 xcp-dev python[808]: twitchAPI.type.TwitchBackendException: can't connect
Jul 17 02:01:25 xcp-dev python[808]: Connection to chat websocket lost and unable to reestablish connection!
Jul 17 02:01:25 xcp-dev python[808]: Traceback (most recent call last):
Jul 17 02:01:25 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/aiohttp/client_ws.py", line 244, in receive
Jul 17 02:01:25 xcp-dev python[808]:     msg = await self._reader.read()
Jul 17 02:01:25 xcp-dev python[808]:           ^^^^^^^^^^^^^^^^^^^^^^^^^
Jul 17 02:01:25 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/aiohttp/streams.py", line 681, in read
Jul 17 02:01:25 xcp-dev python[808]:     return await super().read()
Jul 17 02:01:25 xcp-dev python[808]:            ^^^^^^^^^^^^^^^^^^^^
Jul 17 02:01:25 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/aiohttp/streams.py", line 640, in read
Jul 17 02:01:25 xcp-dev python[808]:     await self._waiter
Jul 17 02:01:25 xcp-dev python[808]: asyncio.exceptions.CancelledError
Jul 17 02:01:25 xcp-dev python[808]: During handling of the above exception, another exception occurred:
Jul 17 02:01:25 xcp-dev python[808]: Traceback (most recent call last):
Jul 17 02:01:25 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/twitchAPI/chat/__init__.py", line 913, in __task_receive
Jul 17 02:01:25 xcp-dev python[808]:     message = await self.__connection.receive(timeout=receive_timeout)
Jul 17 02:01:25 xcp-dev python[808]:               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jul 17 02:01:25 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/aiohttp/client_ws.py", line 243, in receive
Jul 17 02:01:25 xcp-dev python[808]:     async with async_timeout.timeout(timeout or self._receive_timeout):
Jul 17 02:01:25 xcp-dev python[808]:   File "/usr/lib/python3.11/asyncio/timeouts.py", line 98, in __aexit__
Jul 17 02:01:25 xcp-dev python[808]:     raise TimeoutError
Jul 17 02:01:25 xcp-dev python[808]: TimeoutError
Jul 17 02:01:25 xcp-dev python[808]: During handling of the above exception, another exception occurred:
Jul 17 02:01:25 xcp-dev python[808]: Traceback (most recent call last):
Jul 17 02:01:25 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/twitchAPI/chat/__init__.py", line 918, in __task_receive
Jul 17 02:01:25 xcp-dev python[808]:     await self._handle_base_reconnect()
Jul 17 02:01:25 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/twitchAPI/chat/__init__.py", line 956, in _handle_base_reconnect
Jul 17 02:01:25 xcp-dev python[808]:     await self.__connect(is_startup=False)
Jul 17 02:01:25 xcp-dev python[808]:   File "/home/andy/software/pytwitchchatlogger/venv/lib/python3.11/site-packages/twitchAPI/chat/__init__.py", line 866, in __connect
Jul 17 02:01:25 xcp-dev python[808]:     raise TwitchBackendException('can\'t connect')
Jul 17 02:01:25 xcp-dev python[808]: twitchAPI.type.TwitchBackendException: can't connect
Jul 18 10:22:28 xcp-dev systemd[751]: Stopping pytwitchchatlogger.service...
```

There apparently isn't a way that this exception could be caught so that the client can tell the library to initiate a re-connect. 

Posting here at Lena's request (from Discord)

**Comment by iturdikulov** [comment]
I think this is possible, need to increase `reconnect_delay_steps` https://github.com/Teekeks/pyTwitchAPI/blob/2b90657491108cec172578afa0967c88282eddd0/twitchAPI/chat/__init__.py#L589C9-L589C81

Modify `Chat` instance, some like this:
```python
chat = await Chat(twitch)
reconnect_steps = [i*i for i in range(100)]
print(f"Hours to reconnect: {sum(reconnect_steps) // 3600}")  # replace to logger
chat.reconnect_delay_steps = reconnect_steps
```

I not verified this, but it can work.

**Comment by Teekeks** [comment]
correct, thats what that var is for

**Comment by ToastedWookiee** [comment]
Have this same issue with `eventsub` the exception is never called and handled in `except TwitchBackendException:` in my try block. Stopped the websocket and api using a `KeyboardInterrupt` to call the `.stop()` and `.close()` functions.

```12:40:41 DEBUG    got session keep alive
12:41:01 WARNING  keepalive missed, connection lost. reconnecting...
12:41:01 DEBUG    reconnecting using wss://eventsub.wss.twitch.tv/ws...
12:41:01 WARNING  connection attempt failed, retry in 0 seconds...
12:41:01 WARNING  connection attempt failed, retry in 1 seconds...
12:41:02 WARNING  connection attempt failed, retry in 2 seconds...
12:41:04 WARNING  connection attempt failed, retry in 4 seconds...
12:41:08 WARNING  connection attempt failed, retry in 8 seconds...
12:41:11 DEBUG    websocket is closing
12:41:16 WARNING  connection attempt failed, retry in 16 seconds...
12:41:32 WARNING  connection attempt failed, retry in 32 seconds...
12:42:10 WARNING  connection attempt failed, retry in 64 seconds...
12:43:14 WARNING  connection attempt failed, retry in 128 seconds...

13:39:21 DEBUG    stopping websocket EventSub...
13:39:22 ERROR    Task exception was never retrieved
future: <Task finished name='Task-17' coro=<EventSubWebsocket._task_reconnect_handler() done, defined at /home/scott/.local/lib/python3.12/site-packages/twitchAPI/eventsub/websocket.py:303> exception=TwitchBackendException("can't connect to EventSub websocket wss://eventsub.wss.twitch.tv/ws")>
Traceback (most recent call last):
  File "/home/scott/.local/lib/python3.12/site-packages/twitchAPI/eventsub/websocket.py", line 312, in _task_reconnect_handler
    await self._connect(is_startup=False)
  File "/home/scott/.local/lib/python3.12/site-packages/twitchAPI/eventsub/websocket.py", line 275, in _connect
    raise TwitchBackendException(f'can\'t connect to EventSub websocket {self.connection_url}')
twitchAPI.type.TwitchBackendException: can't connect to EventSub websocket wss://eventsub.wss.twitch.tv/ws
```

Got around it by modifying `.reconnect_delay_steps` from https://github.com/Teekeks/pyTwitchAPI/issues/318#issuecomment-2378449708 so far, at least it keeps retrying after when it would do the 128sec wait.

**Comment by Teekeks** [comment]
thinking more about this: it does not really make sense to stop retrying if the error is a connection problem.

I will probably change it to retry indefinitely with the period of the longest entry in that list

**Comment by altskop** [comment]
Ran into this issue earlier myself and have been looking into it, and I think there's more going on here.

To give some context, my bot would lose connection and enter the reconnect loop, but it would never successfully reconnect; I have raised the `reconnect_delay_steps` var up to 6 hours and it did not seem to help. The output, invariably, would be something like this:
```
2026-02-17 06:21:17,484 - twitchAPI.chat - WARNING - connection attempt failed, retry in 60s...
2026-02-17 06:22:17,486 - twitchAPI.chat - WARNING - connection attempt failed, retry in 60s...
2026-02-17 06:23:17,489 - twitchAPI.chat - WARNING - connection attempt failed, retry in 60s...
2026-02-17 06:27:17,500 - twitchAPI.chat - WARNING - connection attempt failed, retry in 60s...
2026-02-17 06:28:17,502 - twitchAPI.chat - WARNING - connection attempt failed, retry in 60s...
```

Looking at `chat/__init__.py`, the failing line that is throwing the exception seems to be L875. I changed the catch block in my environment to capture the exception details for further investigation:
```
try:
    self.__connection = await self._session.ws_connect(self.connection_url)
except Exception as e:
    self.logger.exception(e)
    self.logger.warning(f'connection attempt failed, retry in {self.reconnect_delay_steps[retry]}s...')
```
With this change, I was able to capture the underlying exception:
```
2026-02-17 14:23:17,339 - twitchAPI.chat - ERROR - cannot schedule new futures after shutdown
Traceback (most recent call last):
  File "E:\Studio\Code\Twitch API\pytwitch-sandbox\venv\Lib\site-packages\twitchAPI\chat\__init__.py", line 875, in __connect
    self.__connection = await self._session.ws_connect(self.connection_url)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Studio\Code\Twitch API\pytwitch-sandbox\venv\Lib\site-packages\aiohttp\client.py", line 1076, in _ws_connect
    resp = await self.request(
           ^^^^^^^^^^^^^^^^^^^
  File "E:\Studio\Code\Twitch API\pytwitch-sandbox\venv\Lib\site-packages\aiohttp\client.py", line 779, in _request
    resp = await handler(req)
           ^^^^^^^^^^^^^^^^^^
  File "E:\Studio\Code\Twitch API\pytwitch-sandbox\venv\Lib\site-packages\aiohttp\client.py", line 734, in _connect_and_send_request
    conn = await self._connector.connect(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Studio\Code\Twitch API\pytwitch-sandbox\venv\Lib\site-packages\aiohttp\connector.py", line 672, in connect
    proto = await self._create_connection(req, traces, timeout)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Studio\Code\Twitch API\pytwitch-sandbox\venv\Lib\site-packages\aiohttp\connector.py", line 1239, in _create_connection
    _, proto = await self._create_direct_connection(req, traces, timeout)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Studio\Code\Twitch API\pytwitch-sandbox\venv\Lib\site-packages\aiohttp\connector.py", line 1562, in _create_direct_connection
    hosts = await self._resolve_host(host, port, traces=traces)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Studio\Code\Twitch API\pytwitch-sandbox\venv\Lib\site-packages\aiohttp\connector.py", line 1178, in _resolve_host
    return await asyncio.shield(resolved_host_task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Studio\Code\Twitch API\pytwitch-sandbox\venv\Lib\site-packages\aiohttp\connector.py", line 1209, in _resolve_host_with_throttle
    addrs = await self._resolver.resolve(host, port, family=self._family)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\Studio\Code\Twitch API\pytwitch-sandbox\venv\Lib\site-packages\aiohttp\resolver.py", line 40, in resolve
    infos = await self._loop.getaddrinfo(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Max\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 867, in getaddrinfo
    return await self.run_in_executor(
                 ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Max\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 829, in run_in_executor
    executor.submit(func, *args), loop=self)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Max\AppData\Local\Programs\Python\Python311\Lib\concurrent\futures\thread.py", line 167, in submit
    raise RuntimeError('cannot schedule new futures after shutdown')
RuntimeError: cannot schedule new futures after shutdown
2026-02-17 14:23:17,339 - twitchAPI.chat - WARNING - connection attempt failed, retry in 16s...
```
This runtime error seems to indicate that the executor aiohttp was using has been shut down, and is likely the root cause of every reconnect attempt failing.
Now, this is reproducible with a very minimal chatbot example on Python 3.11. The easiest way to reproduce the issue is to briefly cut your connection after starting the bot and observe the errors you get.

I have tested this on Python 3.10 and 3.14, where I ran into the same issue.
However, when testing on Python 3.8 the behavior is as expected and reconnects work fine.
I double checked the front page of the repo and the tags indicate that all these versions should be supported.

So to summarize, it seems that the reconnect logic is broken in Python 3.10+. Admittedly, I have not tested on 3.9, so it's possible the problem was introduced even then. 

For now, the workaround seems to be to stick with Python 3.8. I attempted to install the same `aiohttp` version in 3.11 as I had in 3.8, but the problem persisted, implying it probably originates from changes in the threading implementation in newer Python versions. I'm not entirely sure what causes this effect yet.

Looking at @adhawkins 's and @ToastedWookiee 's logs, they both seem to be running Python 3.11 and 3.12 respectively, which is likely the root cause in their case too.
